### PR TITLE
Issues/74 ai

### PR DESCRIPTION
**Pull request by AurumTheEnd**
I chose the **Enum polymorphism** instead of the **dyn Trait** one to avoid having to change `game.players` into a `Vec<Box<dyn ParticipantTrait>` monstrosity. This solution required the least changes to code outside `gamestate/players/mod.rs`.

Inspired by https://www.mattkennedy.io/blog/rust_polymorphism/. If you have ideas of how to reduce code duplication in `gamestate/players/mod.rs` and between `gamestate/players/player.rs` & `gamestate/players/ai.rs`, I'm all ears.

Addresses the domain part of #74.

### CODE REVIEW

**Comment by AurumTheEnd** [comment]
@mazanec202 ideas about code deduplication?

**Comment by AurumTheEnd** [comment]
@mazanec202 please comment if attaching `ai_turn` at the end of previous player turns is a good idea or not

**Comment by AurumTheEnd** [comment]
@mazanec202 please comment if using thread::sleep for simulation AI decision time is a good idea or not

**Comment by holubond** [comment]
@AurumTheEnd 

>I chose the Enum polymorphism instead of the dyn Trait one to avoid having to change game.players into a Vec<Box<dyn ParticipantTrait> monstrosity. This solution required the least changes to code outside gamestate/players/mod.rs.

>@mazanec202 ideas about code deduplication?

I'd consider simply adding a field `is_ai` to the `Player` struct. the difference between Human and AI is not that significant to use polymorphism (hence the massive code duplicity).

Alternatively I'd use [impl Trait](https://doc.rust-lang.org/rust-by-example/trait/impl_trait.html), it does not require using `Box` and provides compile-time safety. It might be possible to use some default method implementations to reduce the code duplicity.

> @mazanec202 please comment if attaching `ai_turn` at the end of previous player turns is a good idea or not

The optimal solution would be IMO to lean towards an event-driven architecture of the game logic, but we don't really have time for such modification. Your solution is fine as long as:
 - game does not go on hold when AI is the first player to play in a game (not sure right now)
 - AI can play after an AI played (which I believe is possible now)

> @mazanec202 please comment if using thread::sleep for simulation AI decision time is a good idea or not

We definitely want a non-blocking sleep (no sleep would be better than a blocking one), I'd recommend https://actix.rs/actix/actix/clock/fn.sleep.html - we already have the actix runtime, so all we need to do is make ancestor functions async.

**Comment by AurumTheEnd** [comment]
> @AurumTheEnd
> 
> > I chose the Enum polymorphism instead of the dyn Trait one to avoid having to change game.players into a Vec<Box monstrosity. This solution required the least changes to code outside gamestate/players/mod.rs.
> 
> > @mazanec202 ideas about code deduplication?
> 
> I'd consider simply adding a field `is_ai` to the `Player` struct. the difference between Human and AI is not that significant to use polymorphism (hence the massive code duplicity).

Huh, yeah, missed such a simple solution... will fix.

> 
> Alternatively I'd use [impl Trait](https://doc.rust-lang.org/rust-by-example/trait/impl_trait.html), it does not require using `Box` and provides compile-time safety. It might be possible to use some default method implementations to reduce the code duplicity.

You cannot put `impl Trait` into a Vector (game.players) because size isn't known at compile time.

> 
> > @mazanec202 please comment if attaching `ai_turn` at the end of previous player turns is a good idea or not
> 
> The optimal solution would be IMO to lean towards an event-driven architecture of the game logic, but we don't really have time for such modification. Your solution is fine as long as:
> 
>     * game does not go on hold when AI is the first player to play in a game (not sure right now)

Currently probably holds. Will fix.

> 
>     * AI can play after an AI played (which I believe is possible now)

Currently should be fine, should result in a recursive chaining of `play_card` -> `play_card` -> `play_card` (sometimes substitute for `draw_cards`) until a human player has the turn. As long as we dont overflow the recursion stack this should be fine... aside from when:
 - player has to draw cards, that information is send at the very end of `draw_cards`
 - the next player is AI, recursive calling occurs
 - recursive calling finishes, the original player now gets their cards via the HTTP API
> 
> 
> > @mazanec202 please comment if using thread::sleep for simulation AI decision time is a good idea or not
> 
> We definitely want a non-blocking sleep (no sleep would be better than a blocking one), I'd recommend https://actix.rs/actix/actix/clock/fn.sleep.html - we already have the actix runtime, so all we need to do is make ancestor functions async.

I'll look into it.

**Comment by AurumTheEnd** [comment]
@Karolina42 `todo!` ChainedAiError handling needs to be properly addressed in backend-util.

**Comment by AurumTheEnd** [comment]
image
@mazanec202 This is unfixable. `Line 44` executes current card drawing, and all subsequent AI plays within it, before the original player gets his `HttpResponse::Ok()` with his drawn cards back.

**Comment by AurumTheEnd** [comment]
> > > @mazanec202 please comment if using thread::sleep for simulation AI decision time is a good idea or not
> > 
> > 
> > We definitely want a non-blocking sleep (no sleep would be better than a blocking one), I'd recommend https://actix.rs/actix/actix/clock/fn.sleep.html - we already have the actix runtime, so all we need to do is make ancestor functions async.
> 
> I'll look into it.

@mazanec202 It requires a change of unimagible proportions, including test functions , game start, playcard, drawcards,... I am not in the mood for such large refactoring. Blocking sleep or no sleep then?

**Comment by holubond** [comment]
> image @mazanec202 This is unfixable. `Line 44` executes current card drawing, and all subsequent AI plays within it, before the original player gets his `HttpResponse::Ok()` with his drawn cards back.

I see, the correct behavior would require an AI to have its own thread (or use a separate thread for the game), or at least to send cards to a player via websocket. We don't have time for that. I believe we have two options:
 - send the cards and a next player info (line 54) to the player via websocket before the AI plays (otherwise the player receives an incorrect `next` player, which brings inconsistency to the frontend)
 - don't use AI at all

I'll start a discussion

File: backend/src/gamestate/game.rs [hunk 1]
```diff
@@ -289,17 +326,8 @@ impl Game {
...
-            player_name.clone(),
-            WSMsg::draw(
-                player_name,
-                self.get_current_player().unwrap().name(),
-                draw_count,
-            ),
-        );
-
-        Ok(drawn_cards)
+        assert_eq!(draw_count, drawn_cards.len());
```
Comment (holubond) [inline backend/src/gamestate/game.rs]: Toto patří do testů, tady nám to může potenciálně shodit server

Comment (AurumTheEnd) [reply backend/src/gamestate/game.rs]: @mazanec202 A my to na serveru běžíme v debug módu, ne v releasu?

Comment (holubond) [reply backend/src/gamestate/game.rs]: Obávám se, že Rustu je to jedno...

> Assertions are always checked in both debug and release builds, and cannot be disabled.

 https://doc.rust-lang.org/std/macro.assert.html

`debug_assert_eq!` bude to, co hledáme

Comment (AurumTheEnd) [reply backend/src/gamestate/game.rs]: Fixed in rebase.

**Pull request merged**